services/nginz/third_party/nginx-zauth-module/zauth_module.c

```diff
@@ -59,7 +66,15 @@ static ngx_command_t zauth_commands [] = {
         , NGX_HTTP_LOC_CONF | NGX_CONF_TAKE1
         , ngx_conf_set_flag_slot
         , NGX_HTTP_LOC_CONF_OFFSET
-        , offsetof (ZauthLocationConf, toggle)
+        , offsetof (ZauthLocationConf, zauth)
+        , NULL
+        }
+
+       , { ngx_string ("oauth")
```

Comment:
I'm personally no fan of suddenly having the nginz zauth module now also handle oauth as well, and pushing more of this logic into nginz.

Is there a way for backend services to do this verification?

If there is not, we should at least namespace these options, and make sure this behaviour is properly documented.

Reply:
The intention was to have wire-server handle OAuth requests and do all the verifications, and that's how it's implemented.
At the same time, we want the same routes for both OAuth and ZAuth.
Therefore, nginz needs to let OAuth requests through to wire-server. The changes that are made to nginz in this PR are IMO the most minimal invasive to accomplish this.

I agree with the documentation.

I don't know what you mean with the namespaces?

Reply:
What I meant to say was i don't know / understand what the new `oauth yes` directive in the zauth nginx module is doing, what it does with these requests if "it let's these requests through" and "wire-server handles oauth requests". Why does zauth need to handle these specifically at all? Can't we just bypass it for these paths?

Reply:
The problem is, that it is exactly the same path. No way to decide if this is an OAuth or ZAuth request other than by trying to parse it as ZAuth and see if it succeeds or not. Or do you have a better idea of how to achieve this?

Reply:
So wherever `oauth yes` is set, this simply means "if authentication (seemingly) fails, simply let through anyways?

Do we do more zauth-style validation inside each service as well, or what happens if I access one of these `oauth yes` endpoints?

Reply:
@flokli i agree with your sentiment, but maybe a little documentation will make this a tolerable design.  something like this:

> zauth paths can be set up to additionally support oauth: for routes configured accordingly, the `Authorization` header is processed for zauth credentials, and if that fails, instead of responding with 403, the `Authorization` header is passed to the service intact for oauth processing.

while i'm writing this, i'm wondering if

- (1) we can do anything about the risk of nginz expecting brig to handle security, and brig expecting nginz to do that.  some ad-hoc script that reads both the swagger of the public api and the nginz yaml, and breaks CI if anything goes through by accident?
- (2) should we even name that oauth, or should we call it "custom-service-auth" or something?
- (3) where should we document this?

Reply:
If `oauth yes` is set and authentication fails, nginz lets the request through under the following conditions:
- The request has an Authorization header with a bearer token
- All the other Zauth headers (e.g. Z-User etc.) are emptied
- The token (from the Authorization header) is written to the `Z-OAuth` header

In wire-server, we check if there are proper ZAuth headers (they can only be present if nginz successfully verified the request as a zauth request).

If not, wire-server looks at the `Z-OAuth` header and does the proper OAuth verification including checking the correct sope. If that succeeds, the request handler is called.